File: kotlin/android/segment.kt
...
import com.segment.analytics.Analytics
import com.segment.analytics.Properties
import com.segment.analytics.Traits
...

private class SegmentDestination(context: Context) {

    private val applicationContext = context.applicationContext

    private lateinit var segment: Analytics

    private fun hashMapToProperties(map: Map<String, *>): Properties {
        val properties = Properties()
        for ((k, v) in map) {
        properties.putValue(k, v)
        }
        return properties
    }

    private fun hashMapToTraits(map: Map<String, *>): Traits {
        val traits = Traits()
        for ((k, v) in map) {
        traits.putValue(k, v)
        }
        return traits
    }

    override fun make(env: AvoEnv, apiKey: String?) {
        segment = Analytics.Builder(applicationContext, apiKey)
        .trackApplicationLifecycleEvents()
        .recordScreenViews()
        .build()
    }

    override fun logEvent(eventName: String, eventProperties: Map<String, *>) {
        segment.track(eventName, hashMapToProperties(eventProperties))
    }

    override fun setUserProperties(userId: String, userProperties: Map<String, *) {
        segment.identify(hashMapToTraits(userProperties))
    }

    override fun logPage(pageName: String, eventProperties: Map<String, *>) {
        segment.screen(pageName, hashMapToProperties(eventProperties))
    }

    override fun revenue(amount: Double, eventProperties: Map<String, *>) {
        segment.track("User spent: $amount", hashMapToProperties(eventProperties))
    }

    override fun identify(userId: String) {
        segment.identify(userId)
    }

    override fun unidentify() {
        segment.reset()
    }
}

val segmentDestinationInterface = SegmentDestination(context.applicationContext)


AvoImpl(...,
      myDestination = segmentDestinationInterface)

File: typescript/node/segment.ts
// Install the official Segment Node.js SDK: npm i --save analytics-node
import Analytics from "analytics-node"; // Tested with version 6.2.0

// Get types from the DefenitelyTyped library: npm install --save-dev @types/analytics-node

// Import the Avo Codegen file from the location you've chosen
import { CustomDestination } from "../path/to/Avo";

let analytics: Analytics | undefined;

const segmentDestinationInterface: CustomDestination = {
  make: (_env, apiKey) => {
    analytics = new Analytics(apiKey, { flushAt: 1 });
  },

  identify: (userId) => analytics?.identify({ userId }),

  logEvent: (userId, eventName, eventProperties) =>
    analytics?.track({ userId, event: eventName, properties: eventProperties }),

  setUserProperties: (userId, userProperties) =>
    analytics?.identify({ userId, traits: userProperties }),

  logPage: (userId, pageName, eventProperties) =>
    analytics?.page({ userId, name: pageName, properties: eventProperties }),

  revenue: (userId, amount, eventProperties) =>
    analytics?.track({
      userId,
      event: "Revenue",
      properties: { ...eventProperties, amount },
    }),

  // The following methods are used for group analytics and you might not need them. Learn more about group analytics here: https://www.avo.app/docs/data-design/groups
  addCurrentUserToGroup: (userId, _groupType, groupId) =>
    analytics?.group({ userId, groupId }),

  setGroupProperties: (userId, _groupType, groupId, groupProperties) =>
    analytics?.group({ userId, groupId, traits: groupProperties }),

  // Not supported by the Segment SDK
  // logEventWithGroups: (eventName, eventProperties, groupTypesToGroupIds) => {},
};
